[quant] Add missing functions and methods for channelwise quantization

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24936 [quant] per channel quantization support
* #24935 [quant] Per Channel quantization APIs
* **#24934 [quant] Add missing functions and methods for channelwise quantization**

1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

dskhudia committed:
[quant] Add missing functions and methods for channelwise quantization

1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

dskhudia committed:
Update on "[quant] Add missing functions and methods for channelwise quantization"


1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

raghuramank100 commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -640,6 +640,8 @@ class CAFFE2_API Tensor {
   Tensor dequantize() const;
   double q_scale() const;
   int64_t q_zero_point() const;
+  Tensor q_scales() const;
```
Lets call it something more distinctive: q_per_axis_scale and q_per_axis_zero_point? The current names differ by one letter and are unclear

jerryzh168 replied:
should be per_channel I think

raghuramank100 commented on torch/_tensor_docs.py:
```diff
@@ -1898,6 +1898,22 @@ def callable(a, b) -> number
+
+add_docstr_all('q_zero_points',
+               r"""
+q_zero_points() -> Tensor
```
Please change the names to be consistent (see comment on q_scales and q_zero_points)

raghuramank100 reviewed (changes requested):
Some comments

dskhudia committed:
Update on "[quant] Add missing functions and methods for channelwise quantization"


1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

dskhudia commented:
> Some comments

Addressed comments.

dskhudia committed:
Update on "[quant] Add missing functions and methods for channelwise quantization"


1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

raghuramank100 commented on test/test_quantized_tensor.py:
```diff
@@ -52,13 +52,15 @@ def test_qtensor(self):
         self.assertEqual(qr.item(), 15)
         # we can also print a qtensor
-        self.assertEqual(str(qr),
+        self.assertEqual(' '.join(str(qr).split()),
```
Please add a test for q_per_channel_scales() and q_per_channel_zero_points()

dskhudia replied:
Test for q_per_channel_scales() and q_per_channel_zero_points() are added in the followup diff. Please checkout 'test_per_channel_qtensor_creation' here: https://github.com/pytorch/pytorch/pull/24935/files 

quantize_per_channel and dequanize() already exists. Please check test_qtensor_per_channel_affine in this same file (test/test_quantized_tensor.py).

raghuramank100 reviewed (changes requested):
Please add tests for per channel quant, thanks

dskhudia commented:
@raghuramank100 : The tests exist. Please see my comment.

dskhudia committed:
Update on "[quant] Add missing functions and methods for channelwise quantization"


1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)

dskhudia committed:
Update on "[quant] Add missing functions and methods for channelwise quantization"


1) Functions and methods to get scales and zero_points for channelwise quantization were missing. Adding these.
2) Correctly print quantized tensors for channelwise quantization.

Differential Revision: [D16659715](https://our.internmc.facebook.com/intern/diff/D16659715/)
